### PR TITLE
[FEATURE] [SERVER] chat field support server

### PR DESCRIPTION
**Pull request by burtenshaw**
# Description
This PR adds support for a chat field to the server.

**Type of change**

- New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

Tests are updated

**Checklist**

- I added relevant documentation
- I followed the style guidelines of this project
- I did a self-review of my code
- I made corresponding changes to the documentation
- I confirm My changes generate no new warnings
- I have added tests that prove my fix is effective or that my feature works
- I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

### CODE REVIEW

File: argilla-server/src/argilla_server/search_engine/commons.py [hunk 1]
```diff
@@ -150,6 +150,15 @@ def es_mapping_for_field(field: Field) -> dict:
...
+    if field_type == FieldType.chat:
+        es_field = {
+            "type": "object",
+            "properties": {
+                "content": {"type": "text"},
+                "role": {"type": "keyword"},
+            },
+        }
+        return {es_field_for_record_field(field.name): es_field}
     elif field_type == FieldType.image:
```
Comment (davidberenstein1957) [inline argilla-server/src/argilla_server/search_engine/commons.py]:
```suggestion
    if field.is_text:
        return {es_field_for_record_field(field.name): {"type": "text"}}
    if field.is_chat:
        es_field = {
            "type": "object",
            "properties": {
                "content": {"type": "text"},
                "role": {"type": "keyword"},
            },
        }
        return {es_field_for_record_field(field.name): es_field}
    elif field.is_image:
```

File: argilla-server/src/argilla_server/validators/records.py [hunk 2]
```diff
@@ -108,6 +109,36 @@ def _validate_image_field(self, field_name: str, field_value: Union[str, None])
...
 
+    def _validate_chat_fields(self, dataset: Dataset, fields: Dict[str, str]) -> None:
+        for field in filter(lambda field: field.is_chat, dataset.fields):
+            self._validate_chat_field(field.name, fields.get(field.name))
+
+    def _validate_chat_field(self, field_name: str, field_value: Union[str, None]) -> None:
+        if field_value is None:
+            return
+
+        if len(field_value) > 5000:
```
Comment (davidberenstein1957) [inline argilla-server/src/argilla_server/validators/records.py]: I think we should create a constant for this

Comment (davidberenstein1957) [reply argilla-server/src/argilla_server/validators/records.py]: Also, is the number based on something specific or just intuitively chosen?

File: argilla-server/src/argilla_server/validators/records.py [hunk 3]
```diff
@@ -108,6 +109,36 @@ def _validate_image_field(self, field_name: str, field_value: Union[str, None])
...
+            raise UnprocessableEntityError(
+                f"chat field {field_name!r} value is exceeding the maximum length of 5000 characters"
+            )
+
+        if not isinstance(field_value, list):
+            raise UnprocessableEntityError(f"chat field {field_name!r} value must be a list of dictionaries")
+
+        for i, value in enumerate(field_value):
```
Comment (davidberenstein1957) [inline argilla-server/src/argilla_server/validators/records.py:129]: I think we should also validate if we don't pass anything extra.

File: argilla-server/src/argilla_server/validators/records.py [hunk 4]
```diff
@@ -108,6 +109,36 @@ def _validate_image_field(self, field_name: str, field_value: Union[str, None])
...
+                    f"chat field {field_name!r} value must be a list of dictionaries. Found a non-dictionary value at index {i}"
+                )
+            if "content" not in value:
+                raise UnprocessableEntityError(
+                    f"chat field {field_name!r} value must be a list of dictionaries with a 'content' key. Missing 'content' key at index {i}"
+                )
+            if "role" not in value:
+                raise UnprocessableEntityError(
+                    f"chat field {field_name!r} value must be a list of dictionaries with a 'role' key. Missing 'role' key at index {i}"
+                )
```
Comment (davidberenstein1957) [inline argilla-server/src/argilla_server/validators/records.py]: Similar to the mention earlier about `index` not always being helpful directly. Perhaps we can show the record or the value?

File: argilla-server/tests/unit/api/handlers/v1/datasets/records/records_bulk/test_create_dataset_records_bulk.py [hunk 5]
```diff
@@ -412,3 +413,129 @@ async def test_create_dataset_records_bulk_with_image_field_data_url_exceeding_m
...
+        )
+
+        assert response.status_code == 422
+        assert response.json() == {
+            "detail": f"record at position 0 is not valid because chat field 'chat' value at position 0 is missing the 'role' key",
+        }
+
+        assert (await db.execute(select(func.count(Record.id)))).scalar_one() == 0
+
+    async def test_create_dataset_records_bulk_with_chat_field_without_content_key(
```
Comment (davidberenstein1957) [inline argilla-server/tests/unit/api/handlers/v1/datasets/records/records_bulk/test_create_dataset_records_bulk.py:538]: we should test for toooo long content too

File: argilla-server/tests/unit/database/models/test_field_model.py [hunk 6]
```diff
@@ -29,3 +29,8 @@ def test_is_image_property(self):
         assert Field(settings={"type": FieldType.image}).is_image == True
         assert Field(settings={"type": FieldType.text}).is_image == False
         assert Field(settings={}).is_image == False
+
+    def test_is_chat_property(self):
```
Comment (davidberenstein1957) [inline argilla-server/tests/unit/database/models/test_field_model.py:35]: to be complete, should we add image here too? Also to the test above.

File: argilla-server/src/argilla_server/validators/records.py (hunk 3 above)
Comment (burtenshaw) [reply argilla-server/src/argilla_server/validators/records.py:129]: Sorry, I don't quite get this. We should validate that field_value does contains only content and role keys?

Comment (davidberenstein1957) [reply argilla-server/src/argilla_server/validators/records.py:129]: @burtenshaw what happens if we pass.
```
{
    "role": "user",
    "content": "Hello!",
    "extra": "Hello!"
 }
```

File: argilla-server/src/argilla_server/validators/records.py [hunk 7]
```diff
@@ -108,6 +110,36 @@ def _validate_image_field(self, field_name: str, field_value: Union[str, None])
...
+        for field in filter(lambda field: field.is_chat, dataset.fields):
+            self._validate_chat_field(field.name, fields.get(field.name))
+
+    def _validate_chat_field(self, field_name: str, field_value: Union[str, None]) -> None:
+        if field_value is None:
+            return
+
+        if len(field_value) > CHAT_FIELD_MAX_LENGTH:
+            raise UnprocessableEntityError(
+                f"chat field {field_name!r} value is exceeding the maximum length of 5000 characters"
```
Comment (davidberenstein1957) [inline argilla-server/src/argilla_server/validators/records.py]:
```suggestion
                f"chat field {field_name!r} value is exceeding the maximum length of {CHAT_FIELD_MAX_LENGTH} characters"
```

File: argilla-server/src/argilla_server/validators/records.py [hunk 8]
```diff
@@ -108,6 +110,36 @@ def _validate_image_field(self, field_name: str, field_value: Union[str, None])
...
+            )
+
+        if not isinstance(field_value, list):
+            raise UnprocessableEntityError(f"chat field {field_name!r} value must be a list of dictionaries")
```
Comment (davidberenstein1957) [inline argilla-server/src/argilla_server/validators/records.py:127]: should we also output and example of the list of dictionaries?

File: argilla-server/src/argilla_server/validators/records.py [hunk 9]
```diff
@@ -108,6 +110,36 @@ def _validate_image_field(self, field_name: str, field_value: Union[str, None])
...
+                )
+            if "content" not in value:
+                raise UnprocessableEntityError(
+                    f"chat field {field_name!r} value must be a list of dictionaries with a 'content' key. Missing 'content' key at index {i}. Value is {value}"
+                )
+            if "role" not in value:
+                raise UnprocessableEntityError(
+                    f"chat field {field_name!r} value must be a list of dictionaries with a 'role' key. Missing 'role' key at index {i}. Value is {value}"
+                )
+
```
Comment (davidberenstein1957) [inline argilla-server/src/argilla_server/validators/records.py:142]: I think we might want to check if they don't add other random keys?

File: argilla-server/tests/unit/api/handlers/v1/datasets/fields/test_list_dataset_fields.py [hunk 10]
```diff
@@ -19,7 +19,7 @@
 
 from argilla_server.enums import FieldType
 
-from tests.factories import DatasetFactory, ImageFieldFactory
+from tests.factories import DatasetFactory, ImageFieldFactory, ChatFieldFactory
```
Comment (davidberenstein1957) [inline argilla-server/tests/unit/api/handlers/v1/datasets/fields/test_list_dataset_fields.py:22]: Unrelated, but why don't we have tests for TextFields here?

File: argilla-server/src/argilla_server/validators/records.py [hunk 11]
```diff
@@ -40,6 +40,7 @@
     "image/svg",
     "image/webp",
 ]
+CHAT_FIELD_MAX_LENGTH = 500
```
Comment (davidberenstein1957) [inline argilla-server/src/argilla_server/validators/records.py:43]: should we add these to general constants.py?

File: argilla-server/tests/unit/api/handlers/v1/datasets/records/records_bulk/test_create_dataset_records_bulk.py [hunk 12]
```diff
@@ -412,3 +413,154 @@ async def test_create_dataset_records_bulk_with_image_field_data_url_exceeding_m
...
+                    },
+                ],
+            },
+        )
+
+        assert response.status_code == 422
+        assert "is not valid" in response.json()["detail"]
+        assert (await db.execute(select(func.count(Record.id)))).scalar_one() == 0
+
+    async def test_create_dataset_records_bulk_with_chat_field_without_content_key(
```
Comment (davidberenstein1957) [inline argilla-server/tests/unit/api/handlers/v1/datasets/records/records_bulk/test_create_dataset_records_bulk.py:538]: I think we might need and extra check to validate they don't add extra keys. Also, I think the we might simplify the key-check tests using the parameterize over a set of keys within the json payload.

File: argilla-server/tests/unit/api/handlers/v1/test_datasets.py [hunk 13]
```diff
@@ -2065,6 +2065,7 @@ async def test_create_dataset_records_with_wrong_value_field(
         )
 
         assert response.status_code == 422
+        print(response.json())
```
Comment (davidberenstein1957) [inline argilla-server/tests/unit/api/handlers/v1/test_datasets.py]:
```suggestion
```

File: argilla-server/src/argilla_server/validators/records.py (hunk 8 above)
Comment (burtenshaw) [reply argilla-server/src/argilla_server/validators/records.py:127]: I think that would be excessive.

File: argilla-server/src/argilla_server/validators/records.py (hunk 9 above)
Comment (burtenshaw) [reply argilla-server/src/argilla_server/validators/records.py:142]: Some datasets use other keys like 'timestamp' or 'turn': https://huggingface.co/datasets/HannahRoseKirk/prism-alignment

I think this is ok to support and not something we could easily validate on a first iteration.

File: argilla-server/tests/unit/api/handlers/v1/datasets/fields/test_list_dataset_fields.py (hunk 10 above)
Comment (burtenshaw) [reply argilla-server/tests/unit/api/handlers/v1/datasets/fields/test_list_dataset_fields.py:22]: I'm not sure why they're not here. `TextFieldFactory` is tested in `test_datasets.py`.

File: argilla-server/tests/unit/api/handlers/v1/datasets/records/records_bulk/test_create_dataset_records_bulk.py (hunk 12 above)
Comment (burtenshaw) [reply argilla-server/tests/unit/api/handlers/v1/datasets/records/records_bulk/test_create_dataset_records_bulk.py:538]: I responded above to extra keys. Open to a PR on the feature branch for improved validation/testing.

File: argilla-server/src/argilla_server/validators/records.py (hunk 11 above)
Comment (burtenshaw) [reply argilla-server/src/argilla_server/validators/records.py:43]: Personally, I find local constants easier to read and maintain.

**Review by davidberenstein1957 — APPROVED** [review]

File: argilla-server/src/argilla_server/validators/records.py (hunk 11 above)
Comment (davidberenstein1957) [reply argilla-server/src/argilla_server/validators/records.py:43]: Me too normally but otherwise it ends up being fragmented, right?

**Pull request merged**